internal/balancer/balanceraggregator/balanceraggregator.go

```diff
@@ -0,0 +1,168 @@
+}
+
+// Parent is a balancer.ClientConn that can also receive
+// child state updates.
+type Parent interface {
+	balancer.ClientConn
+	UpdateChildState(childStates []ChildState)
+}
+
+// Build returns a new BalancerAggregator.
+func Build(parent Parent, opts balancer.BuildOptions) *BalancerAggregator {
+	return &BalancerAggregator{
+		parent:   parent,
+		bOpts:    opts,
+		children: resolver.NewEndpointMap(),
+	}
+}
+
+// BalancerAggregator is a balancer that wraps child balancers. It creates a
+// child balancer with child config for every Endpoint received. It updates the
+// child states on any update from parent or child.
+type BalancerAggregator struct {
+	parent Parent
+	bOpts  balancer.BuildOptions
+
+	children *resolver.EndpointMap
+
+	inhibitChildUpdates bool
+}
+
+func (ba *BalancerAggregator) UpdateClientConnState(state balancer.ClientConnState) error {
+	endpointSet := resolver.NewEndpointMap()
+	ba.inhibitChildUpdates = true
+	// Update/Create new children.
+	for _, endpoint := range state.ResolverState.Endpoints {
+		endpointSet.Set(endpoint, nil)
+		var bal *balancerWrapper
+		if child, ok := ba.children.Get(endpoint); ok {
+			bal = child.(*balancerWrapper)
+		} else {
+			bal = &balancerWrapper{
+				endpoint:   endpoint,
+				ClientConn: ba.parent,
+				ba:         ba,
+			}
+			bal.Balancer = gracefulswitch.NewBalancer(bal, ba.bOpts)
+			ba.children.Set(endpoint, bal)
+		}
+		if err := bal.UpdateClientConnState(balancer.ClientConnState{
+			BalancerConfig: state.BalancerConfig,
+			ResolverState: resolver.State{
+				Endpoints:  []resolver.Endpoint{endpoint},
+				Attributes: state.ResolverState.Attributes,
+			},
+		}); err != nil {
+			return fmt.Errorf("error updating child balancer: %v", err)
+		}
+	}
+	// Delete old children that are no longer present.
+	for _, e := range ba.children.Keys() {
+		child, _ := ba.children.Get(e)
+		bal := child.(balancer.Balancer)
+		if _, ok := endpointSet.Get(e); !ok {
+			bal.Close()
+			ba.children.Delete(e)
+		}
+	}
+	ba.inhibitChildUpdates = false
+	ba.updateChildStates()
+	return nil
+}
+
+func (ba *BalancerAggregator) ResolverError(err error) {
+	ba.inhibitChildUpdates = true
+	for _, child := range ba.children.Values() {
+		bal := child.(balancer.Balancer)
+		bal.ResolverError(err)
+	}
+	ba.inhibitChildUpdates = false
+	ba.updateChildStates()
+}
+
+func (ba *BalancerAggregator) UpdateSubConnState(sc balancer.SubConn, state balancer.SubConnState) {
+	// UpdateSubConnState is deprecated.
+}
+
+func (ba *BalancerAggregator) Close() {
+	for _, child := range ba.children.Values() {
+		bal := child.(balancer.Balancer)
+		bal.Close()
+	}
+}
+
+func (ba *BalancerAggregator) updateChildStates() {
+	if ba.inhibitChildUpdates {
+		return
+	}
+
+	childUpdates := make([]ChildState, ba.children.Len())
+	for _, child := range ba.children.Values() {
+		bw := child.(*balancerWrapper)
+		childUpdates = append(childUpdates, ChildState{
+			Endpoint: bw.endpoint,
+			State:    bw.state,
+		})
```

Comment:
Maybe store a ChildState instead of the fields separately?

Reply:
Done. Feels a little weird since I set 50% of it in constructor and the rest in UpdateState, but it cleans this line up, and I guess it's doing that here anyway, just adds a nesting layer to the data structure that aligns to what it'll send upward anyway. Thanks.